### PR TITLE
[RLlib] Fix IMPALA/APPO learning behavior: Fix EnvRunner sync bug, GPU loader thread, enable local learner w/ GPU.

### PR DESCRIPTION
**Pull request by sven1977**
Enhance IMPALA/APPO learning behavior:
* Fix EnvRunner sync bug (when running with lots of EnvRunners, only those few that - by chance - return samples in the same call to `training_step` as IMPALA/APPO performs the Learner update, get weight-synched. All others (even though they also contributed to the train batch) stay behind, leading to a very large off-policyness.
* GPU loader thread: Disable loading to the GPU on the NumpyToTensor connector. This load process should be handled entirely by the n GPU loader threads.
* Enable local learner w/ GPU: When using a local Learner (num_learners=0), it should already use the GPU, if available.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by aslonnie — COMMENTED** [review]
seems that it has merge conflicts now?

File: rllib/algorithms/impala/impala.py [hunk 1]
```diff
@@ -721,7 +702,11 @@ def training_step(self) -> ResultDict:
                     )
                 if not do_async_updates:
                     learner_results = [learner_results]
+
                 for results_from_n_learners in learner_results:
+                    if not results_from_n_learners[0]:
```
Comment (simonsays1980) [inline rllib/algorithms/impala/impala.py:707]: Didn't we continue here before?

File: rllib/algorithms/impala/impala.py [hunk 2]
```diff
@@ -942,12 +932,19 @@ def default_resource_request(
...
+                            cf.num_cpus_per_learner if cf.num_learners == 0 else 0,
+                        )
+                        + cf.num_aggregation_workers
+                    ),
+                    # Use n GPUs if we have a local Learner (num_learners=0).
+                    "GPU": (
+                        (cf.num_gpus_per_learner if cf.num_learners == 0 else 0)
+                        if cf.enable_rl_module_and_learner
+                        else (0 if cf._fake_gpus else cf.num_gpus)
+                    ),
```
Comment (simonsays1980) [inline rllib/algorithms/impala/impala.py:947]: This sets the GPUs to reserve to zero when using remote learners. Is this defined elsewhere?

**Review by simonsays1980 — COMMENTED** [review]
LGTM. One small question to resource request.

**Review by simonsays1980 — APPROVED** [review]

File: rllib/algorithms/impala/impala.py (hunk 2 above)
Comment (sven1977) [reply rllib/algorithms/impala/impala.py:947]: which would be correct. If we use remote learners, the main process doesn't need any GPU anymore.

File: rllib/algorithms/impala/impala.py (hunk 1 above)
Comment (sven1977) [reply rllib/algorithms/impala/impala.py:707]: No, we sent the empty results into the logger and tried to extract the model state from it, which wouldn't succeed. So it was pretty much a waste of time :)

**Pull request merged**